### PR TITLE
feature/5353-NR-Timeout-Changes

### PR DESCRIPTION
**Pull request by bcdevlucas**
*Issue #:* /bcgov/entity#5353

*Description of changes:*
See commit messages.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namerequest license (Apache 2.0).

### CODE REVIEW

**Comment by severinbeauvais** [comment]
Could you please update the issue number and add a few details in the description block above?

**Comment by severinbeauvais** [comment]
Also please assign yourself to this PR.

**Comment by severinbeauvais** [comment]
Also could you please rebase your branch to the latest master so that we only see YOUR changes (not 26 commits! haha).

File: client/src/App.vue
Comment (severinbeauvais) [inline client/src/App.vue:106]: These are not all async methods. (This doesn't hurt but it's misleading.)

Async methods are those that can be blocked on something, like a network call.

Also on line 122 below.

Comment (severinbeauvais) [reply client/src/App.vue:106]: (FWIW, we often need async/await in the unit tests to let Vuetify do some things before we verify stuff. In my experience we haven't needed this in the actual code.)

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (severinbeauvais) [inline client/src/components/session-timer/session-timer-mixin.ts:21]: This is a bit odd. Should this code be waiting for the dispatch call to finish? Or we don't care what happens (ie, empty then function)?

I would expect to see something like this (from onTimerModelSessionExtended):
```
expirationFn: async () => {
   await this.$store.dispatch(types.SHOW_NR_SESSION_EXPIRY_MODAL)
}```

Note that the arrow function is async, and it awaits the dispatch call.

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (severinbeauvais) [inline client/src/components/session-timer/session-timer-mixin.ts:40]: As far as I can tell, createAndStartTimer is not declared as async, so returning it isn't async, and making THIS method async doesn't do anything.

File: client/src/components/session-timer/countdown-timer-display.vue
Comment (severinbeauvais) [inline client/src/components/session-timer/countdown-timer-display.vue:5]: This is cool :) but why didn't you use the Vuetify progress component? Eg,
https://vuetifyjs.com/en/components/progress-circular/#rotate

File: client/src/components/payment/upgrade-modal.vue
Comment (severinbeauvais) [inline client/src/components/payment/upgrade-modal.vue:71]: Is this just debugging code? If not, why does this exist?

Comment (severinbeauvais) [reply client/src/components/payment/upgrade-modal.vue:71]: Ditto timerName below...?

File: client/src/components/new-request/main-container.vue
Comment (severinbeauvais) [inline client/src/components/new-request/main-container.vue:14]: Our UX/UI designers have become careful about our colour usage, so if possible use named colours (or Material named colours) (or variables from theme.scss) instead of hex colours here. (Same reason we use constants in code.)

File: client/src/components/new-request/main-container.vue
Comment (severinbeauvais) [inline client/src/components/new-request/main-container.vue:54]: As far as I know, `this.$set` is to add properties to a reactive object.
If I understand correctly, this is not the case here... `displayTimer` is a class property and simply updating it should update the prop in the countdown-timer element.
If I misunderstand, please explain, thanks!

File: client/src/components/payment/payment-modal.vue
Comment (severinbeauvais) [inline client/src/components/payment/payment-modal.vue:63]: Why implement props this way instead of with `@Prop` decorator+method?

File: client/src/components/payment/upgrade-modal.vue
Comment (severinbeauvais) [inline client/src/components/payment/upgrade-modal.vue:2]: Keep this in mind for future dialogs: you should pass a prop to them to "attach" to. In the parent you would pass a selector to something. This allows a dialog to be attached to the parent, so you can unit test it! (Otherwise it's attached to the DOM root which you can't see in the unit test.)

Also it would be great if components had a id or class, for testing purposes!

File: client/src/store/new-request-module.ts
Comment (severinbeauvais) [inline client/src/store/new-request-module.ts:612]: Looks OK but a couple of comments:
- you can use TS conditional chaining here, eg. `if (this.analysisJSON?.issues ...`
- let's start adding return types to methods/getters, to make it easier to review and implement new code

**Review by severinbeauvais — CHANGES_REQUESTED** [review]
See comments above.

File: client/src/App.vue
Comment (bcdevlucas) [reply client/src/App.vue:106]: Actions are async in Vue and can be awaited eg. resetAnalyzeName. Probably shouldn't have awaited those mutations.

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (bcdevlucas) [reply client/src/components/session-timer/session-timer-mixin.ts:21]: We don't care what happens. I am aware of how to use async and ES fat arrow functions as well...
Awaiting the display of that NR session expiry modal blocks I/O and slows things down and I explicitly don't want that to occur. When the session expiry modal displays, a new timer starts for 45 seconds (or whatever that window is set to).

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (bcdevlucas) [reply client/src/components/session-timer/session-timer-mixin.ts:40]: Again, in Vue, actions are async. createAndStartTimer is an action.

File: client/src/components/session-timer/countdown-timer-display.vue
Comment (bcdevlucas) [reply client/src/components/session-timer/countdown-timer-display.vue:5]: I wanted something to look specifically the way I wanted to, and to keep it lightweight, CSS only. Stroke is an easy property to work with.

File: client/src/components/payment/upgrade-modal.vue
Comment (bcdevlucas) [reply client/src/components/payment/upgrade-modal.vue:71]: That just allows you to optionally display timers on the update, receipt, and re-apply pages should you want to. Lorna and I were discussing whether or not we wanted timers on those, but then she went on holiday.

File: client/src/components/new-request/main-container.vue
Comment (bcdevlucas) [reply client/src/components/new-request/main-container.vue:14]: Busy cranking out fixes, I can update that, just trying to get things working here. I need to do a css clean up in a couple areas related to payments. Noted regarding the colors. The colors I have used are the same ones you use for fonts and buttons.

File: client/src/components/payment/payment-modal.vue
Comment (bcdevlucas) [reply client/src/components/payment/payment-modal.vue:63]: @Prop decorator doesn't appear to cover all use cases that you can do with standard Vue syntax. Also that library used to add those @Prop are a 3rd party library not core Vue functionality IIRC.

File: client/src/store/new-request-module.ts
Comment (bcdevlucas) [reply client/src/store/new-request-module.ts:612]: Merge fragment. Not sure why git detected that as a change.

File: client/src/components/new-request/main-container.vue
Comment (bcdevlucas) [reply client/src/components/new-request/main-container.vue:54]: this.$set from within a component allows you set data from within a 'method' function of a Vue component. This would be more clear using standard Vue syntax as opposed to the component decorators.
https://stackoverflow.com/questions/51296654/vuejs-set-data-in-methods-function

File: client/src/App.vue
Comment (severinbeauvais) [reply client/src/App.vue:106]: Are you sure? I see some async actions but not "resetAnalyzeName"...?

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (severinbeauvais) [reply client/src/components/session-timer/session-timer-mixin.ts:21]: I know you are aware of async and arrow functions. I have seen lots of your code use them! 😃 Sorry, I thought this might be an oversight or incomplete code.

Why bother having a "then" block? I think you can just leave it off. Either way, a comment would explain this and prevent a future developer from scratching their head.

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (severinbeauvais) [reply client/src/components/session-timer/session-timer-mixin.ts:40]: I'm genuinely trying to understand this.

Ref: https://vuex.vuejs.org/guide/actions.html#dispatching-actions-in-components

It looks like actions _can_ be async, but it looks like this one isn't.
```
export const createAndStartTimer = ({ commit }, opts: TimerOptions) => {
  commit(types.CREATE_AND_START_TIMER, opts)
}
```

Let me know if I misunderstand.

File: client/src/components/session-timer/countdown-timer-display.vue
Comment (severinbeauvais) [reply client/src/components/session-timer/countdown-timer-display.vue:5]: Sounds good.

File: client/src/components/payment/upgrade-modal.vue
Comment (severinbeauvais) [reply client/src/components/payment/upgrade-modal.vue:71]: Where is the conditional? As far as I can tell, this always returns false.

How about adding a comment indicating this is for possible future use?

File: client/src/components/new-request/main-container.vue
Comment (severinbeauvais) [reply client/src/components/new-request/main-container.vue:14]: It's OK for now.

Note that we have tickets asking to update some of those text/font/button/icon styles.

So it will be a game of catch-up. Later!

File: client/src/components/new-request/main-container.vue
Comment (severinbeauvais) [reply client/src/components/new-request/main-container.vue:54]: Thanks for the explanation. That SO post is specifically for adding/changing an object's properties, as explained here:
https://www.telerik.com/blogs/so-what-actually-is-vue-set

As for an object watcher, it only fires when the object itself is reassigned. To watch for object property changes you have to use this:
```
@Watch('myObject', { deep: true })
```

Leave it for now but I think it's tech debt related to mixed use of object style vs component style. AFAIK the Entity UIs have not needed to use `this.$set` for any reason.

File: client/src/components/payment/payment-modal.vue
Comment (severinbeauvais) [reply client/src/components/payment/payment-modal.vue:63]: I'm curious about that but leave it for now.

BTW, you already import that library to provide `@Prop`... it's "vue-property-decorator'".

File: client/src/store/new-request-module.ts
Comment (severinbeauvais) [reply client/src/store/new-request-module.ts:612]: OK. This is another reason why a rebase would have been great before review :)

File: client/src/App.vue
Comment (bcdevlucas) [reply client/src/App.vue:106]: Whether you label them as async or not, an action is dispatched, and as result is inherently async.
Typescript / linting doesn't pick up on it...
https://medium.com/@bogna.ka/everyday-life-in-vue-world-vuex-actions-asynchronous-nature-ad154d246e4b

File: client/src/components/payment/upgrade-modal.vue
Comment (severinbeauvais) [reply client/src/components/payment/upgrade-modal.vue:2]: I created a new tech debt ticket for this:
https://app.zenhub.com/workspaces/entity-5bf2f2164b5806bc2bf60531/issues/bcgov/entity/5492

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (bcdevlucas) [reply client/src/components/session-timer/session-timer-mixin.ts:21]: The then block prevents a linting issue with the current settings that does not allow an unawaited or then'ed call on an async function. I don't want to await those calls, so I just added the empty then block to get around the warning... I left notes in the code in most of the places I did that: 
```
// Don't await this request, that way there's no lag, fire it off async and don't block I/O	      // Don't await this request, that way there's no lag, fire it off async and don't block I/O
      // The empty then clause just prevents a linting issue that warns when you don't	      // The empty then clause just prevents a linting issue that warns when you don't
      // await an async function which is not an issue, since we don't want to block I/O
```

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (bcdevlucas) [reply client/src/components/session-timer/session-timer-mixin.ts:40]: Dispatching an action always returns a promise. I guess I didn't declare the actions async when i did the VX timer store module, but yes, dispatching those will also be async and returns a promise. Vuex does it under the hood when the action is registered.
```
export const createAndStartTimer = ({ commit }, opts: TimerOptions)
```

File: client/src/App.vue
Comment (severinbeauvais) [reply client/src/App.vue:106]: Oh wow. Thanks for the link. It would be nice if Vue fixed their docs then!!!

OK, as a **future** exercise, action method declarations should be updated to return an explicit promise type, so a developer doesn't have to guess whether the return type is void, any, or async!

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (severinbeauvais) [reply client/src/components/session-timer/session-timer-mixin.ts:21]: OK, cool.

File: client/src/components/session-timer/session-timer-mixin.ts
Comment (severinbeauvais) [reply client/src/components/session-timer/session-timer-mixin.ts:40]: It's weird Vue-ism and new to me.

This will all be a lot easier to sort through when types are implement globally (in another ticket).

OK for now.

**Review by severinbeauvais — APPROVED** [review]

**Comment by severinbeauvais** [comment]
Please wait for Cam's review before merging.

File: client/src/components/payment/payment-modal.vue
Comment (cameron-freshworks) [inline client/src/components/payment/payment-modal.vue:28]: Nothing major but i think within the Vue template itself "" is the standard and elsewhere in this project.

Comment (bcdevlucas) [reply client/src/components/payment/payment-modal.vue:28]: True.

**Review by cameron-freshworks — APPROVED** [review]
Quite a bit i don't understand but my comments were minor. LGTM, especially if it unblocks you. 

@bcdevlucas Just let me know when your g2g and we'll get er in 👍

**Comment by bcdevlucas** [comment]
@cameron-freshworks Thanks.

**Pull request merged**
